### PR TITLE
Add LTV

### PR DESCRIPTION
**Pull request by evgeny-stakewise**

### CODE REVIEW

File: src/ltv/tasks.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        return None
+
+    ipfs_hash = event['args']['rewardsIpfsHash']
+    rewards_root = event['args']['rewardsRoot']
+    rewards_data = await ipfs_fetch_client.fetch_json(ipfs_hash)
+    vault_to_harvest_params = get_vault_to_harvest_params(
+        reward_snapshot=RewardSnapshot.from_dict(rewards_data),
+        rewards_root=rewards_root,
+    )
+    return vault_to_harvest_params.get(vault)
```
Comment (tsudmi) [inline src/ltv/tasks.py:136]: Harvest params can also be fetched from the subgraph:
rewardsRoot, proofReward, proofUnlockedMevReward, proof

File: src/ltv/settings.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# IPFS
+IPFS_FETCH_ENDPOINTS = config(
+    'IPFS_FETCH_ENDPOINTS',
+    cast=Csv(),
+    default=','.join(
+        ['https://stakewise-v3.infura-ipfs.io', 'https://gateway.pinata.cloud', 'https://ipfs.io']
+    ),
+)
+
+IPFS_FETCH_CLIENT_TIMEOUT = config('IPFS_FETCH_CLIENT_TIMEOUT', default='60', cast=int)
```
Comment (tsudmi) [inline src/ltv/settings.py:25]: Please clean up unused variables

File: deploy/update-ltv.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+apiVersion: batch/v1
```
Comment (tsudmi) [inline deploy/update-ltv.yaml:1]: Lets not add the kubernetes deployments here. It's better to have a periodic tasks helm chart here: https://github.com/stakewise/helm-charts/

File: README.md
Comment (tsudmi) [inline README.md:1]: The readme file is empty

File: .env.example [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+PRIVATE_KEY=0x2......
+TARGET_CHAIN=23
+TARGET_ADDRESS=0xbd335c16c94be8c4dd073ae376ddf78bec1858df
+PRICE_FEED=0xba74737a078c05500dd98c970909e4a3b90c35c6
+PRICE_FEED_SENDER=0xf7d4e7273e5015c96728a6b02f31c505ee184603
+
+EXECUTION_ENDPOINT=https://execution
+VAULT_USER_LTV_TRACKER_CONTRACT_ADDRESS=0x...
+GRAPH_API_URL=https://graph
+VAULT=0x...
```
Comment (tsudmi) [inline .env.example:12]: Please clean up the variables here. We have MAINNET_RPC_URL and EXECUTION_ENDPOINT that are the same. Also separate by blocks the configuration for different periodic tasks

File: deploy/update-price.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+apiVersion: batch/v1
```
Comment (tsudmi) [inline deploy/update-price.yaml:1]: same as above

File: src/ltv/clients.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    HOT_WALLET_PRIVATE_KEY,
+)
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+hot_wallet_account = Account().from_key(HOT_WALLET_PRIVATE_KEY)
+execution_client = Web3(Web3.HTTPProvider(EXECUTION_ENDPOINT))
+execution_client.middleware_onion.add(construct_sign_and_send_raw_middleware(hot_wallet_account))
+logger.info('hot wallet address: %s', hot_wallet_account.address)
```
Comment (tsudmi) [inline src/ltv/clients.py:22]:
```suggestion
logger.info('Wallet address: %s', hot_wallet_account.address)
```

File: src/ltv/contracts.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        while to_block >= from_block:
+            events = event.get_logs(
+                fromBlock=BlockNumber(max(to_block - blocks_range, from_block)),
+                toBlock=to_block,
+                argument_filters=argument_filters,
+            )
+            if events:
+                return events[-1]
+            to_block = BlockNumber(to_block - blocks_range - 1)
+        return None
```
Comment (tsudmi) [inline src/ltv/contracts.py:57]: Looks like events are not used.

File: src/ltv/typings.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+                vault_reward = OwnMevVaultReward(
+                    vault=vault['vault'],
+                    consensus_reward=consensus_reward,
+                    proof=vault['proof'],
+                    unlocked_mev_reward=vault['unlocked_mev_reward'],
+                    withdrawals=withdrawals,
+                )
+
+            vault_rewards.append(vault_reward)
+        return RewardSnapshot(vaults=vault_rewards)
```
Comment (tsudmi) [inline src/ltv/typings.py:106]: As we use the subgraph calls, are these dataclasses actually used?

File: src/get_ltv.py
Comment (tsudmi) [inline src/get_ltv.py:1]: How is this script used?

Comment (evgeny-stakewise) [reply src/get_ltv.py:1]: `python src/get_ltv.py`

File: deploy/update-ltv.yaml (hunk 3 above)
Comment (evgeny-stakewise) [reply deploy/update-ltv.yaml:1]: FYI: there is manifest.yml file in multichain repo, contains kubernetes stuff.

File: src/ltv/tasks.py [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    logger.info('Tx confirmed')
+
+    # Get LTV after update
+    ltv = vault_user_ltv_tracker_contract.get_vault_max_ltv(VAULT, harvest_params)
+    logger.info('LTV after update: %s', Decimal(ltv) / WAD)
+
+    logger.info('Completed')
+
+
+def get_vault_max_ltv() -> None:
```
Comment (cyc60) [inline src/ltv/tasks.py:75]: Not used, if it a debug func pls add more details in docsting.

Comment (cyc60) [inline src/ltv/tasks.py:75]:
```suggestion
def pring_vault_max_ltv() -> None:
```

File: src/ltv/clients.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+
+def get_graph_client() -> gql.Client:
+    if GRAPH_API_URL is None:
+        raise ValueError('GRAPH_API_URL must be set')
+
+    if GRAPH_API_TIMEOUT is None:
```
[the remark this answers is not in the document]
Comment (evgeny-stakewise) [reply src/ltv/clients.py:29]: There is default value 10

**Review by cyc60 — APPROVED** [review]

File: deploy/update-ltv.yaml (hunk 3 above)
Comment (tsudmi) [reply deploy/update-ltv.yaml]: We should remove that from there and move to helm charts

**Pull request merged**